keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+  Existing deployments will need to remove `descheduler.alpha.kubernetes.io/request-evict-only`
+  annotation to utilize the new evacuation API.
+  Otherwise, the eviction API will be used as a fallback.
+- The evacuation API expects the evacuation request to be reconciled during each descheduling cycle.
+  Given each cycle starts from scratch the descheduler needs to first list
+  all evacuation requests and sort the pods accordingly in each plugin.
+  Other components may create an evacuation request as well.
+  The descheduler needs to take this into account.
+  E.g. by creating a snapshot (per cycle, per strategy) to avoid inconsistencies.
+  An evacuation request may exist without a corresponding pod
+  before the evacuation controller garbage collects the request.
+- The descheduler needs to keep a cache of evacution API requests (CRs)
+  and remove itself from finalizers if the request is no longer needed.
+  E.g. `evacuation.coordination.k8s.io/instigator_descheduler.sigs.k8s.io` as a finalizer.
+- Read the eviction status from the [evacuation CR status](https://github.com/kubernetes/enhancements/pull/4565).
+
+#### Code changes
+
+The code responsible for invoking the eviction API is located
+under `sigs.k8s.io/descheduler/pkg/descheduler/evictions` under private `evictPod` function.
+The function can be generalized into a generic interface:
+
+```go
+type PodEvacuator interface {
+    EvacutePod(ctx context.Context, pod *v1.Pod) error
+}
+```
+
+to provide the [descheduling framework](https://github.com/kubernetes-sigs/descheduler/pull/1372)
+a generic way of evicting/evacuating pods.
+The new interface can be used for implementing various testing scenarios.
+Or turned into a plugin if needed.
+
+#### Metrics
+
+A new metric `pod_evacuations` for counting the number of evacuated pods is introduced.
+The descheduler will not observe whether a pod was ultimately evicted.
+Only provide a summary about how many pods were requested to be evacuated.
+If needed, another metric `pod_evacuations_in_progress` for counting
+the number of evacutions that are reconciled can be introduced.
+
+### Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding
+to implement the design.
+
+* The evacution API proposal mentions the evacuation instigator should remove
+  its intent when the evacuation is no longer necessary.
+  Should each descheduling cycle reset all the eviction requests
+  or wait until some of the eviction requests were completed
+  to free a "bucket" for new evictions?
+  **Answer**: The first implementation will not reset any evacuation request.
+  The descheduler will account for existing requests and update the internal counters accordingally.
+  In the future a mechanism for deleting too old evacuation requests can be introduced.
+  I.e. based on a new `--max-evacuation-request-ttl` option.
+* The evacution API proposal mentions more than a one entity can request an eviction.
```

Comment:
does the descheduler have a use for a custom value of `.spec.progressDeadlineSeconds`? We are thinking of using a unified value (`1800 (30m)`) in all of the components. 

Might be good to clarify deschedulers position on this.

As it currently stands, the entity who creates the Evacuation has a control of this value. The following instigators who do not create the object, do not have the controll.

Reply:
> We are thinking of using a unified value (1800 (30m)) in all of the components.

We can make it the default and configurable if needed. Can an evacuation request be created without `.spec.progressDeadlineSeconds`? I.e. can we expected the default value will get set to a reasonable value/30m?

Reply:
That is a good idea, I will add a default of 30m to the API.